File: client.old/GhostClient/Sprites/Effects/CircleEffect.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Ghost.Sprites.Items;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using Sharp2D;
using Sharp2D.Core.Interfaces;

namespace Ghost.Sprites.Effects
{
    public class CircleEffect : IEffect
    {
        private static Random random = new Random();
        public void Begin(int duration, int size, float x, float y, double rotation)
        {
            int stage1Duration = (int) rotation; //The rotation field is used as the first stage duration
            int stage2Duration = duration - stage1Duration;

            var emittor = new CircleEmittor(stage1Duration, stage2Duration, x, y, size);
            GhostClient.Ghost.CurrentGhostGame.AddLogical(emittor);
        }

        private class CircleEmittor : ILogical
        {
            private const long DURATION = 100;

            private long lastSpawn, start;
            public int radius, stage1Duration, stage2Duration;
            public double startPos;
            public float x, y;

            public bool stage2;
            private int totalSpawn;
            private int cursor = 0;

            public CircleEmittor(int stage1, int stage2, float x, float y, int size)
            {
                this.stage1Duration = stage1;
                this.stage2Duration = stage2;
                this.x = x;
                this.y = y;
                this.startPos = random.Next();
                this.radius = size;
            }

            public void Dispose()
            {
                
            }

            public void Update()
            {
                if (start == 0)
                {
                    start = Environment.TickCount;
                }
                if (Environment.TickCount - lastSpawn < DURATION)
                    return;

                int spawnCount;
                if (stage2)
                {
                    spawnCount = random.Next(10, 30);
                    spawnCount = Math.Min(totalSpawn - cursor, spawnCount);
                    cursor += spawnCount;
                }
                else
                {
                    spawnCount = random.Next(-1, 9);
                }

                for (int i = 0; i < spawnCount; i++)
                {
                    if (!stage2)
                    {
                        var p = new CircleParticle(this) {X = x, Y = y};
                        GhostClient.Ghost.CurrentGhostGame.AddSprite(p);
                    }
                    else
                    {
                        int xd = random.Next();

                        float tempx = (float) (x + (Math.Cos(xd)*radius));
                        float tempy = (float) (y + (Math.Sin(xd)*radius));

                        var p = new CircleParticle(this) {X = tempx, Y = tempy};
                        GhostClient.Ghost.CurrentGhostGame.AddSprite(p);
                    }
                }

                lastSpawn = Environment.TickCount;

                if (!stage2 && Environment.TickCount - start >= stage1Duration)
                {
                    stage2 = true;
                    start = Environment.TickCount;
                    totalSpawn = random.Next(300, 400);
                }
                else if (stage2 && Environment.TickCount - start >= stage2Duration)
                {
                    GhostClient.Ghost.CurrentGhostGame.RemoveLogical(this);
                }
            }
        }

        private class CircleParticle : Entity
        {
            private double SPEED;

            private bool stage2;
            private CircleEmittor emittor;
            private double startPos, counter;

            private Vector2 target = Vector2.Zero;
            private int duration;
            private int start;
            private float sx, sy;
            public CircleParticle(CircleEmittor emittor) : base(0)
            {
                stage2 = emittor.stage2;
                this.emittor = emittor;
                SPEED = (2.0*Math.PI)/(random.Next(emittor.stage1Duration / 2, emittor.stage1Duration)/16.0);
                startPos = emittor.startPos;
                this.BlendMode = BlendState.Additive;
            }

            protected override void OnLoad()
            {
                Texture = GhostClient.Ghost.CurrentGhostGame.Content.Load<Texture2D>("sprites/ball.png");

                Width = Texture.Width;
                Height = Texture.Height;

                NeverClip = true;
                UniformScale = (float)(random.NextDouble() * (0.3f - 0.15f) + 0.15f);
                TintColor = System.Drawing.Color.FromArgb(255, 36, 81, 163);
            }

            protected override void OnUnload()
            {
                emittor = null;
            }

            protected override void OnDispose()
            {
            }

            public override void Update()
            {
                base.Update();

                if (!stage2 && emittor.stage2)
                {
                    stage2 = true;
                }

                if (stage2)
                {
                    TintColor = System.Drawing.Color.FromArgb(255, 170, 19, 27);
                    if (target == Vector2.Zero)
                    {
                        double tx = emittor.x - X;
                        double ty = emittor.y - Y;
                        double angle = Math.Atan2(ty, tx);
                        int dis = random.Next(300, 500);

                        target = new Vector2(X + (dis * (float)Math.Cos(angle)), Y + (dis * (float)Math.Sin(angle)));
                        duration = random.Next(50, 700);
                        start = Environment.TickCount;
                        sx = X;
                        sy = Y;
                    }

                    X = MathUtils.Ease(sx, target.X, duration, Environment.TickCount - start);
                    Y = MathUtils.Ease(sy, target.Y, duration, Environment.TickCount - start);
                    Alpha = MathUtils.Ease(1, 0, duration / 1.5f, Environment.TickCount - start);

                    if (X == target.X && Y == target.Y)
                    {
                        GhostClient.Ghost.CurrentGhostGame.RemoveSprite(this);
                    }
                }
                else
                {
                    counter += SPEED;

                    X = (float) (emittor.x + Math.Cos(startPos + counter)*emittor.radius);
                    Y = (float) (emittor.y + Math.Sin(startPos + counter)*emittor.radius);
                }
            }
        }
    }
}

File: client/core/src/me/eddiep/ghost/client/core/game/sprites/effects/CircleEffect.kt
package me.eddiep.ghost.client.core.game.sprites.effects

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.utils.TimeUtils
import me.eddiep.ghost.client.Ghost
import me.eddiep.ghost.client.core.game.Entity
import me.eddiep.ghost.client.core.logic.Logical
import me.eddiep.ghost.client.core.render.Blend
import me.eddiep.ghost.client.handlers.scenes.SpriteScene
import me.eddiep.ghost.client.utils.Global
import me.eddiep.ghost.client.utils.Vector2f

class CircleEffect : Effect {
    override fun begin(duration: Int, size: Int, x: Float, y: Float, rotation: Double, world: SpriteScene) {
        val stage1Duration = rotation.toLong()
        val stage2Duration = duration - stage1Duration

        val emittor = CircleEmittor(stage1Duration, stage2Duration, x, y, size, world)
        Ghost.getInstance().addLogical(emittor)
    }
}

class CircleEmittor(val stage1: Long, val stage2: Long, val x: Float, val y: Float, val radius: Int, val world: SpriteScene) : Logical {
    val DURATION = 100L
    val startPos = Global.RANDOM.nextInt();

    var start: Long = 0L
    var lastSpawn: Long = 0L
    var isStage2: Boolean = false
    var totalSpawn = 0
    var cursor = 0

    override fun tick() {
        if (start == 0L) {
            start = TimeUtils.millis()
        }
        if (TimeUtils.millis() - lastSpawn < DURATION)
            return

        var spawnCount: Int
        if (isStage2) {
            spawnCount = Global.RANDOM.nextInt(30 - 10) + 10
            spawnCount = Math.min(totalSpawn - cursor, spawnCount)
            cursor += spawnCount
        } else {
            spawnCount = Global.RANDOM.nextInt(9 - -1) + -1
        }

        for (i in 0..spawnCount-1) {
            if (!isStage2) {
                val p = CircleParticle(this)
                p.setCenter(x, y)
                p.setBlend(Blend(GL20.GL_SRC_ALPHA, GL20.GL_ONE));

                world.addEntity(p)
            } else {
                val xd = Global.RANDOM.nextInt()

                val tempx = x + Math.cos(xd.toDouble())*radius
                val tempy = y + Math.sin(xd.toDouble())*radius

                val p = CircleParticle(this)
                p.setCenter(tempx.toFloat(), tempy.toFloat())
                p.setBlend(Blend(GL20.GL_SRC_ALPHA, GL20.GL_ONE));

                world.addEntity(p)
            }
        }

        lastSpawn = TimeUtils.millis()

        if (!isStage2 && TimeUtils.millis() - start >= stage1) {
            isStage2 = true
            start = TimeUtils.millis()
            totalSpawn = Global.rand(300, 400)
        } else if (isStage2 && TimeUtils.millis() - start >= stage2) {
            Ghost.getInstance().removeLogical(this)
        }
    }

    override fun dispose() { }
}

class CircleParticle(val emittor: CircleEmittor): Entity("sprites/ball.png", 0) {
    val speed: Double = (2.0*Math.PI)/(Global.rand((emittor.stage1/2).toInt(), emittor.stage1.toInt())/16.0)
    var counter: Double = 0.0
    var isStage2 = emittor.isStage2
    val startPos = emittor.startPos

    var _target: Vector2f = Vector2f.ZERO
    var duration = 0f
    var start = 0L
    var sx = 0f
    var sy = 0f

    override fun onLoad() {
        super.onLoad()

        setScale(Global.RANDOM.nextFloat()*(0.3f - 0.15f)+0.15f)

        color = if (!isStage2) Color(36 / 255f, 81 / 255f, 163 / 255f, 1f) else Color(170 / 255f, 19 / 255f, 27 / 255f, 1f)
    }

    override fun tick() {
        super.tick()

        if (!isStage2 && emittor.isStage2) {
            isStage2 = true
            color = Color(170 / 255f, 19 / 255f, 27 / 255f, 1f)
        }

        if (isStage2) {
            if (_target == Vector2f.ZERO) {
                val x = centerX
                val y = centerY

                val tx = emittor.x - x
                val ty = emittor.y - y
                val angle = Math.atan2(ty.toDouble(), tx.toDouble())
                val dis = Global.rand(300, 500)

                _target = Vector2f(x + (dis * Math.cos(angle).toFloat()), y + (dis * Math.sin(angle).toFloat()))
                duration = Global.rand(50, 700).toFloat()
                start = TimeUtils.millis()
                sx = x
                sy = y
            }

            val tempx = Entity.ease(sx, _target.x, duration, (TimeUtils.millis() - start).toFloat())
            val tempy = Entity.ease(sy, _target.y, duration, (TimeUtils.millis() - start).toFloat())
            val a = Entity.ease(1f, 0f, duration / 1.5f, (TimeUtils.millis() - start).toFloat())

            setAlpha(a)
            setCenter(tempx, tempy)

            if (centerX == _target.x && centerY == _target.y) {
                parentScene.removeEntity(this)
            }
        } else {
            counter += speed

            val tempx = (emittor.x + Math.cos(startPos + counter)*emittor.radius).toFloat()
            val tempy = (emittor.y + Math.sin(startPos + counter)*emittor.radius).toFloat()

            setCenter(tempx, tempy)
        }
    }
}

